### PR TITLE
Tests for ipv6 support in urlmatch

### PR DESCRIPTION
**Pull request by DerJesko**
https://github.com/qutebrowser/qutebrowser/issues/4006 seems to be working mostly, except a few edge cases. A few of those are tested here. Ideas were taken here:
- https://chromium.googlesource.com/chromium/src.git/+/c008bf657dbc01fadc4ff1ea3df21f9cf237569a%5E%21/
- https://en.wikipedia.org/wiki/IPv6_address
(I fucked up in one of the commit messages, dunno how to change it after the fact)

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/4118)

### CODE REVIEW

**Comment by DerJesko** [comment]
Ok, I see the problem
```
>>> QUrl('http://[fc2e:0e35:bb88::edac]').host()
'fc2e:e35:bb88::edac'
>>> urlmatch.UrlPattern('http://[fc2e:0e35:bb88::edac]')._host
'fc2e:0e35:bb88::edac'
>>> QUrl('http://[fc2e:e35:bb88:0::edac]').host()
'fc2e:e35:bb88::edac'
>>> urlmatch.UrlPattern('http://[fc2e:e35:bb88:0::edac]')._host
'fc2e:e35:bb88:0::edac'
```
`urllib.parse.urlparse` doesn't minimize the representation. `QUrl` does

**Comment by The-Compiler** [comment]
Hmm, I see. I guess `UrlPattern` should minimize the host itself if it's an IPv6 address then. You can probably use the [ipaddress](https://docs.python.org/3.7/library/ipaddress.html) module to do so:

```
>>> parsed = urllib.parse.urlparse('http://[fc2e:e35:bb88:0::edac]')
>>> parsed.netloc
'[fc2e:e35:bb88:0::edac]'
>>> ip = ipaddress.IPv6Address(parsed.netloc[1:-1])
>>> ip
IPv6Address('fc2e:e35:bb88::edac')
>>> str(ip)
'fc2e:e35:bb88::edac'
```

or perhaps use QUrl directly, which makes sure it gets normalized in the same way:

```
>>> QUrl('//' + parsed.netloc).host()
'fc2e:e35:bb88::edac'
```

**Comment by DerJesko** [comment]
Now everything seems to be working. `urllib.parse.urlparse` still excepts invalid ipv6 addresses though

File: qutebrowser/utils/urlmatch.py [hunk 1]
```diff
@@ -177,6 +178,13 @@ def _init_host(self, parsed):
             assert self._host is None
             return
 
+        if not utils.raises(ValueError, ipaddress.IPv6Address, parsed.netloc[1:-1]):
```
Comment (The-Compiler) [inline qutebrowser/utils/urlmatch.py]: I think you could remove this, and instead raise a `ParseError` with `url.errorString()` if `url.isValid()` is false after setting the host.

However, you should probably only do this `if parsed.netloc.startswith('['):` which kinda mirrors the Chromium code.

File: qutebrowser/utils/urlmatch.py [hunk 2]
```diff
@@ -30,6 +30,7 @@
 import urllib.parse
 
 from qutebrowser.utils import utils, qtutils
+from PyQt5.QtCore import QUrl
```
Comment (The-Compiler) [inline qutebrowser/utils/urlmatch.py]: nitpick: Please move this to a separate block before the `from qutebrowser` import, see [pep8](https://www.python.org/dev/peps/pep-0008/#imports).

**Comment by DerJesko** [comment]
Ok, I ran into a small problem with urllib.parse. Real strange behaviour:
```
>>> urllib.parse.urlparse('http://[::1]]').hostname
'::1'
>>> urllib.parse.urlparse('http://[[::1]').hostname
'[::1'
```

File: qutebrowser/utils/urlmatch.py (hunk 2 above)
Comment (DerJesko) [reply qutebrowser/utils/urlmatch.py]: Thanks

**Comment by The-Compiler** [comment]
I opened https://bugs.python.org/issue34360 - feel free to mark the affected tests as expected to fail, with something like

```python
pytest.param('...', marks=pytest.mark.xfail(
            reason="https://bugs.python.org/issue34360"))`
```

File: qutebrowser/utils/urlmatch.py [hunk 3]
```diff
@@ -177,6 +179,15 @@ def _init_host(self, parsed):
             assert self._host is None
             return
 
+        if parsed.netloc.startswith('['):
+            # Using QUrl parsing to minimize ipv6 addresses
+            url = QUrl()
+            url.setHost("[" + parsed.hostname + "]")
```
Comment (The-Compiler) [inline qutebrowser/utils/urlmatch.py]: I don't think the brackets are required with `setHost`?

File: qutebrowser/utils/urlmatch.py [hunk 4]
```diff
@@ -177,6 +179,15 @@ def _init_host(self, parsed):
...
+            if url.host() == "":
+                raise ParseError("Invalid IPv6 URL"+parsed.hostname)
```
Comment (The-Compiler) [inline qutebrowser/utils/urlmatch.py]: Any reason you aren't using `if url.isValid():` and `url.errorString()` here like I said in my earlier comment? :wink:

Comment (DerJesko) [reply qutebrowser/utils/urlmatch.py]: no, I just overlooked it

File: qutebrowser/utils/urlmatch.py (hunk 3 above)
Comment (DerJesko) [reply qutebrowser/utils/urlmatch.py]: This is needed otherwise if the hostname is `[::1]` it isn't declined

Comment (DerJesko) [reply qutebrowser/utils/urlmatch.py]: ok, I read the python bug report you wrote. It seems assuming `urllib.parse.urlparse` works correctly we don't need the brackets

File: qutebrowser/config/configdata.yml [hunk 5]
```diff
@@ -1622,6 +1622,14 @@ url.yank_ignored_parameters:
     - utm_content
   desc: URL parameters to strip with `:yank url`.
 
+url.yank_remove_password:
```
Comment (jgkamat) [inline qutebrowser/config/configdata.yml]: Is this new setting being used anywhere?

File: qutebrowser/config/configdata.py [hunk 6]
```diff
@@ -56,7 +56,7 @@ class Option:
 @attr.s
 class Migrations:
 
-    """Nigrated options in configdata.yml.
+    """Migrated options in configdata.yml.
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.py]: Dang, that's an embarrassing typo :see_no_evil:. Good catch!

File: qutebrowser/config/configdata.yml (hunk 5 above)
Comment (DerJesko) [reply qutebrowser/config/configdata.yml]: no, this was supposed to be for something else. I'll remove this again

**Comment by The-Compiler** [comment]
This LGTM, but given that the main thing in my head are upcoming exams I'm not too confident in my abilities to do a code review :laughing: @jgkamat or @toofar, if you feel like this is okay to merge, go ahead!
